### PR TITLE
Fix Description Length Preventing REST API Key Creation

### LINKED ISSUE #30594: REST API keys are not saved because API key label can be too long
**Issue:**

We have an app where we request our customers to authorize our application in order to use the WooCommerce REST API. In many cases the API keys we got after the authorization were not valid. After some investigation we found out that the API keys are not persisted at all in the Wordpress database during the authorization step.

We don't have an access to the codebase of our customers but after looking at the source code we made few educated guesses and we found the potential issue – [the generation of the API key description](https://github.com/woocommerce/woocommerce/blob/ead815eea471d97335d6233c35cfcfc274686a0e/includes/class-wc-auth.php#L212-L219).

```php
$description = sprintf(
	/* translators: 1: app name 2: scope 3: date 4: time */
	__( '%1$s - API %2$s (created on %3$s at %4$s).', 'woocommerce' ),
	wc_clean( $app_name ),
	$this->get_i18n_scope( $scope ),
	date_i18n( wc_date_format() ),
	date_i18n( wc_time_format() )
);
```

**TLDR:** The generated `$description` is longer than the database constraint `varchar(200)`.

----

**Details:**

We know the API key is not persisted at all because in the [payload for the inserted ID](https://github.com/woocommerce/woocommerce/blob/ead815eea471d97335d6233c35cfcfc274686a0e/includes/class-wc-auth.php#L248) we get `0`.

In our case the `$scope` is `read_write` and the store is using a plugin for translations. Our guess is that the translation is either too long and it doesn't fit the database constraint or the translation is missing and the translation function is returning some kind of error message string that is too long. In either case, _the behaviour is flaky_ when it relies on translation and the outcome is unknown. The workaround we did for one of our customers was to force the English language and the API key got persisted as expected.

**Suggestion:**

Could we drop the translation entirely here? What about if we use a simpler format like `[APP_NAME] - API ([DATE])` or something else that relies on less variables with unknown values. Another option is to truncate the generated description.

### PR DESCRIPTION
**Pull request by jeffstieler**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #30594.

This PR implements the changes to the API Key description value (generated for keys generated via the REST API) suggested in #30594 as well as basic safeguards from causing the same error using the "Add key" UI in settings. It also introduces a user facing error message if the API Key creation fails via the admin interface:

Screen Shot 2021-10-07 at 5 03 11 PM

With this PR, the endpoint generated key description no longer requires translation and will truncate the `app_name` at `170` characters. 

### How to test the changes in this Pull Request:

If your test store uses a self-signed SSL certificate, add this filter:

```php
add_filter( 'http_request_args', function( $args ) {
        $args['sslverify'] = false;
        return $args;
} );
```

1. Go to the following URL for your store: `http://[STORE_DOMAIN]/wc-auth/v1/authorize?app_name=[APP_NAME]&scope=read_write&user_id=1&return_url=[STORE_DOMAIN]&callback_url=[STORE_DOMAIN]` (where [STORE_DOMAIN] is your test store, and [APP_NAME] is at least 200 characters long)
2. Follow the "approve" link
3. Verify that the key is created and the description contains the truncated `app_name`, the text ` - API`, and the current datetime.
4. Go to WooCommerce > Settings > Advanced > REST API
5. Click "Add key"
6. Attempt to enter a description longer than 200 characters (paste in)
7. Click "Generate API key"
8. Verify the key was created and the description truncated at 200 chars
9. Go to WooCommerce > Settings > Advanced > REST API
10. Click "Add key"
11. Use the developer tools to inspect the description field and remove the `maxlength` attribute
12. Paste in a description longer than 200 characters
13. Click "Generate API key"
14. Verify an error message is shown

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

Fix bug when creating REST API keys with very long descriptions.

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by Konamiman** [comment]
Hi @jeffstieler , I'm trying to test this with `curl`:

```
curl -u konamiman:k -v "http://localhost:8034/wc-auth/v1/authorize?app_name=12345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890123456789012345678901234567890ABCDEFG&scope=read&user_id=theuser&return_url=http%3A%2F%2Flocalhost%3A8034%2Fafterauthorize&callback_url=https%3A%2F%2Flocalhost%3A8034%2Fgimmethekey"
```

I get back HTML that once saved to a file and opened looks like this:

image

Then I click the "Approve" button and I get _Error: Invalid nonce verification_ . Am I doing something wrong?

**Comment by jeffstieler** [comment]
Sorry for the trouble @Konamiman - I've updated the test instructions to specify using the browser rather than making a direct GET request and included a filter in case your test store uses a self-signed SSL certificate.

**Review by Konamiman — APPROVED** [review]
